File: service/src/main/java/no/nav/melosys/service/brev/DokumentNavnService.java
package no.nav.melosys.service.brev;

import no.nav.melosys.domain.Aktoer;
import no.nav.melosys.domain.Behandling;
import no.nav.melosys.domain.brev.Mottaker;
import no.nav.melosys.domain.brev.NorskMyndighet;
import no.nav.melosys.domain.kodeverk.Mottakerroller;
import no.nav.melosys.domain.kodeverk.brev.Produserbaredokumenter;
import no.nav.melosys.service.LovvalgsperiodeService;
import no.nav.melosys.service.dokument.BrevmottakerService;
import no.nav.melosys.service.dokument.DokgenService;
import no.nav.melosys.service.dokument.DokumentproduksjonsInfo;
import no.nav.melosys.service.dokument.VedleggTyper;
import org.springframework.stereotype.Service;

import static no.nav.melosys.domain.kodeverk.lovvalgsbestemmelser.Lovvalgbestemmelser_trygdeavtale_uk.UK_ART8_2;

@Service
public class DokumentNavnService {

    private final BrevmottakerService brevmottakerService;
    private final DokgenService dokgenService;
    private final LovvalgsperiodeService lovvalgsperiodeService;

    public DokumentNavnService(BrevmottakerService brevmottakerService, DokgenService dokgenService, LovvalgsperiodeService lovvalgsperiodeService) {
        this.brevmottakerService = brevmottakerService;
        this.dokgenService = dokgenService;
        this.lovvalgsperiodeService = lovvalgsperiodeService;
    }


    public String utledDokumentNavnForProduserbaredokumenterOgMottakerrolle(Behandling behandling, Produserbaredokumenter produserbaredokumenter, Mottakerroller mottakerRolle) {
        if (erTrygdeavtaleVedtaksbrev(produserbaredokumenter)) {
            Mottaker mottaker = brevmottakerService.avklarMottaker(produserbaredokumenter, Mottaker.av(mottakerRolle), behandling);
            return utledDokumentNavnForProduserbaredokumenterOgMottaker(behandling, produserbaredokumenter, mottaker, null);
        }
        return produserbaredokumenter.getBeskrivelse();
    }

    public String utledDokumentNavnForProduserbaredokumenterOgMottaker(Behandling behandling, Produserbaredokumenter produserbaredokumenter, Mottaker mottaker, String standardTekst) {
        if (erTrygdeavtaleVedtaksbrev(produserbaredokumenter)) {
            DokumentproduksjonsInfo dokumentproduksjonsInfo = dokgenService.hentDokumentInfo(produserbaredokumenter);
            return utledDokumentNavn(behandling, dokumentproduksjonsInfo, mottaker);
        }
        return standardTekst != null ? standardTekst : produserbaredokumenter.getBeskrivelse();
    }

    private boolean erTrygdeavtaleVedtaksbrev(Produserbaredokumenter produserbaredokumenter) {
        return produserbaredokumenter.getKode().contains("TRYGDEAVTALE");
    }

    public String utledDokumentNavn(Behandling behandling, DokumentproduksjonsInfo dokumentproduksjonsInfo, Aktoer mottaker) {
        String tittel = utledTittel(behandling, dokumentproduksjonsInfo, mottaker);

        if (behandling.erNyVurdering()) {
            return lagEndringTittel(tittel);
        }

        return tittel;
    }

    private String utledTittel(Behandling behandling, DokumentproduksjonsInfo dokumentproduksjonsInfo, Aktoer mottaker) {
        if (mottaker.erUtenlandskMyndighet()) {
            return dokumentproduksjonsInfo.vedleggsTitler().get(VedleggTyper.ATTEST);
        }

        var vedtaksbrevTittel = dokumentproduksjonsInfo.vedleggsTitler().get(VedleggTyper.VEDTAKSBREV);

        if (NorskMyndighet.SKATTEETATEN.getOrgnr().equals((mottaker.getOrgnr()))) {
            return lagKopiTittel(vedtaksbrevTittel);
        }

        boolean erArtikkel8_2 = lovvalgsperiodeService.hentLovvalgsperiode(behandling.getId()).getBestemmelse() == UK_ART8_2;

        if (mottaker.erBruker()) {
            return erArtikkel8_2 ? vedtaksbrevTittel : dokumentproduksjonsInfo.journalføringsTittel();
        } else {
            return lagKopiTittel(erArtikkel8_2 ? vedtaksbrevTittel : dokumentproduksjonsInfo.journalføringsTittel());
        }
    }

    public String utledDokumentNavn(Behandling behandling, DokumentproduksjonsInfo dokumentproduksjonsInfo, Mottaker mottaker) {
        String tittel = utledTittel(behandling, dokumentproduksjonsInfo, mottaker);

        if (behandling.erNyVurdering()) {
            return lagEndringTittel(tittel);
        }

        return tittel;
    }

    private String utledTittel(Behandling behandling, DokumentproduksjonsInfo dokumentproduksjonsInfo, Mottaker mottaker) {
        if (mottaker.erUtenlandskMyndighet()) {
            return dokumentproduksjonsInfo.vedleggsTitler().get(VedleggTyper.ATTEST);
        }

        var vedtaksbrevTittel = dokumentproduksjonsInfo.vedleggsTitler().get(VedleggTyper.VEDTAKSBREV);

        if (NorskMyndighet.SKATTEETATEN.getOrgnr().equals((mottaker.getOrgnr()))) {
            return lagKopiTittel(vedtaksbrevTittel);
        }

        boolean erArtikkel8_2 = lovvalgsperiodeService.hentLovvalgsperiode(behandling.getId()).getBestemmelse() == UK_ART8_2;

        if (mottaker.getRolle() == Mottakerroller.BRUKER) {
            return erArtikkel8_2 ? vedtaksbrevTittel : dokumentproduksjonsInfo.journalføringsTittel();
        } else {
            return lagKopiTittel(erArtikkel8_2 ? vedtaksbrevTittel : dokumentproduksjonsInfo.journalføringsTittel());
        }
    }

    private String lagKopiTittel(String tittel) {
        return "Kopi av " + org.springframework.util.StringUtils.uncapitalize(tittel);
    }

    private String lagEndringTittel(String tittel) {
        return tittel + " - endring";
    }
}

File: service/src/main/kotlin/no/nav/melosys/service/brev/DokumentNavnService.kt
package no.nav.melosys.service.brev

import no.nav.melosys.domain.Behandling
import no.nav.melosys.domain.brev.Brevbestilling
import no.nav.melosys.domain.brev.FritekstbrevBrevbestilling
import no.nav.melosys.domain.brev.Mottaker
import no.nav.melosys.domain.kodeverk.Mottakerroller
import no.nav.melosys.domain.kodeverk.brev.Produserbaredokumenter
import no.nav.melosys.domain.kodeverk.lovvalgsbestemmelser.trygdeavtale.Lovvalgsbestemmelser_trygdeavtale_gb
import no.nav.melosys.exception.FunksjonellException
import no.nav.melosys.service.LovvalgsperiodeService
import no.nav.melosys.service.dokument.BrevmottakerService
import no.nav.melosys.service.dokument.DokgenService
import no.nav.melosys.service.dokument.DokumentproduksjonsInfo
import no.nav.melosys.service.ftrl.medlemskapsperiode.MedlemskapsperiodeService
import org.springframework.stereotype.Service
import org.springframework.util.StringUtils
import org.springframework.util.StringUtils.hasText


@Service
class DokumentNavnService(
    private val brevmottakerService: BrevmottakerService,
    private val dokgenService: DokgenService,
    private val lovvalgsperiodeService: LovvalgsperiodeService,
    private val medlemskapsperiodeService: MedlemskapsperiodeService
) {
    fun utledDokumentNavnForProduserbartdokumentOgMottakerrolle(
        behandling: Behandling,
        produserbartdokument: Produserbaredokumenter,
        mottakerRolle: Mottakerroller
    ): String =
        utledTittel(
            behandling = behandling,
            produserbartdokument = produserbartdokument,
            dokumentproduksjonsInfo = dokgenService.hentDokumentInfo(produserbartdokument),
            mottakerRolle = mottakerRolle
        )


    fun utledDokumentNavnForProduserbartdokumentOgMottaker(
        behandling: Behandling,
        produserbardokument: Produserbaredokumenter,
        mottaker: Mottaker,
        standardTekst: String
    ): String =
        utledTittel(
            behandling = behandling,
            produserbartdokument = produserbardokument,
            dokumentproduksjonsInfo = dokgenService.hentDokumentInfo(produserbardokument),
            mottaker = mottaker,
            standardTekst = standardTekst
        )

    fun utledTittel(
        behandling: Behandling,
        produserbartdokument: Produserbaredokumenter,
        dokumentproduksjonsInfo: DokumentproduksjonsInfo,
        mottaker: Mottaker? = null,
        mottakerRolle: Mottakerroller? = null,
        brevbestilling: Brevbestilling? = null,
        standardTekst: String? = null
    ): String =
        when {
            produserbartdokument.erTrygdeavtaleVedtak() ->
                utledTittelTrygdeavtale(behandling, produserbartdokument, dokumentproduksjonsInfo, mottaker, mottakerRolle)

            standardTekst != null ->
                standardTekst

            produserbartdokument.erFritekstbrev() ->
                utledTittelFritekstbrev(dokumentproduksjonsInfo, brevbestilling as FritekstbrevBrevbestilling?)

            produserbartdokument == Produserbaredokumenter.VARSELBREV_MANGLENDE_INNBETALING ->
                utledTittelManglendeInnbetaling(behandling.id, dokumentproduksjonsInfo)

            else -> dokumentproduksjonsInfo.journalføringsTittel
        }

    private fun utledTittelManglendeInnbetaling(behandlingID: Long, dokumentproduksjonsInfo: DokumentproduksjonsInfo): String {
        val erFrivilligMedlemskap = medlemskapsperiodeService.hentMedlemskapsperioder(behandlingID).any { it.erFrivillig() }
        return if (erFrivilligMedlemskap) dokumentproduksjonsInfo.alternativTittel else dokumentproduksjonsInfo.journalføringsTittel
    }

    private fun utledTittelFritekstbrev(dokumentproduksjonsInfo: DokumentproduksjonsInfo, brevbestilling: FritekstbrevBrevbestilling?): String {
        if (brevbestilling == null) {
            return dokumentproduksjonsInfo.journalføringsTittel
        }

        val tittel = brevbestilling.dokumentTittel ?: brevbestilling.fritekstTittel ?: ""

        if (!hasText(tittel)) {
            throw FunksjonellException("Tittel til fritekstbrev mangler, behandlingId:" + brevbestilling.behandlingId)
        }

        if ("Request to remain subject to Norwegian legislation" == tittel) {
            return "Søknad om unntak"
        }

        return tittel
    }

    private fun utledTittelTrygdeavtale(
        behandling: Behandling,
        produserbartdokument: Produserbaredokumenter,
        dokumentproduksjonsInfo: DokumentproduksjonsInfo,
        mottaker: Mottaker?,
        mottakerRolle: Mottakerroller?
    ): String {
        val mottakerAvBrev = mottaker ?: brevmottakerService.avklarMottaker(produserbartdokument, Mottaker.medRolle(mottakerRolle), behandling)
        val tittel = utledTittelTrygdeavtale(behandling.id, dokumentproduksjonsInfo, mottakerAvBrev)
        return if (behandling.erNyVurdering()) lagEndringTittel(tittel) else tittel
    }

    private fun utledTittelTrygdeavtale(behandlingID: Long, dokumentproduksjonsInfo: DokumentproduksjonsInfo, mottaker: Mottaker): String {
        if (mottaker.erUtenlandskMyndighet()) {
            return dokumentproduksjonsInfo.attestTittel ?: throw FunksjonellException("Forventer at trygdeavtale-brev har attest-tittel")
        }

        val erArtikkel8_2 = lovvalgsperiodeService.hentLovvalgsperiode(behandlingID).bestemmelse == Lovvalgsbestemmelser_trygdeavtale_gb.UK_ART8_2
        val tittel = if (erArtikkel8_2) dokumentproduksjonsInfo.alternativTittel else dokumentproduksjonsInfo.journalføringsTittel

        if (mottaker.rolle == Mottakerroller.BRUKER) {
            return tittel
        }

        return lagKopiTittel(tittel)
    }

    private fun lagKopiTittel(tittel: String): String = "Kopi av " + StringUtils.uncapitalize(tittel)

    private fun lagEndringTittel(tittel: String): String = "$tittel - endring"

    private fun Produserbaredokumenter.erFritekstbrev(): Boolean =
        listOf(
            Produserbaredokumenter.FRITEKSTBREV,
            Produserbaredokumenter.GENERELT_FRITEKSTBREV_BRUKER,
            Produserbaredokumenter.GENERELT_FRITEKSTBREV_VIRKSOMHET,
            Produserbaredokumenter.GENERELT_FRITEKSTBREV_ARBEIDSGIVER,
            Produserbaredokumenter.GENERELT_FRITEKSTVEDLEGG,
            Produserbaredokumenter.UTENLANDSK_TRYGDEMYNDIGHET_FRITEKSTBREV
        ).contains(this)

    private fun Produserbaredokumenter.erTrygdeavtaleVedtak(): Boolean =
        listOf(
            Produserbaredokumenter.TRYGDEAVTALE_AU,
            Produserbaredokumenter.TRYGDEAVTALE_GB,
            Produserbaredokumenter.TRYGDEAVTALE_US,
            Produserbaredokumenter.TRYGDEAVTALE_CAN
        ).contains(this)
}
